src/collection-view.js

```diff
@@ -46,33 +50,42 @@ Marionette.CollectionView = Marionette.View.extend({
   },
 
   endBuffering: function() {
+    // Only trigger attach if already shown and attached, otherwise Region#show() handles this.
+    var canTriggerAttach = this._isShown && Marionette.isNodeAttached(this.el);
+    var nestedViews;
+
     this.isBuffering = false;
-    this._triggerBeforeShowBufferedChildren();
 
-    this.attachBuffer(this);
+    if (this._isShown) {
+      this._triggerMethodOnAll(this._bufferedChildren, 'before:show');
+    }
+    if (canTriggerAttach && this.triggerBeforeAttach) {
+      // Recursively retrieve nested views.
+      nestedViews = this._getNestedViews();
+      this._triggerMethodOnAll(nestedViews, 'before:attach');
+    }
 
-    this._triggerShowBufferedChildren();
-    this.initRenderBuffer();
-  },
+    this.attachBuffer(this);
 
-  _triggerBeforeShowBufferedChildren: function() {
+    if (canTriggerAttach && this.triggerAttach) {
```

Comment:
I'm checking that on line 54, because it is possible be "shown" and not attached.  If it's _not_ shown it's certainly not attached, so I'm skipping the `isNodeAttached` check in that case.